src/compiler/crystal/semantic/call_error.cr

```diff
@@ -194,6 +176,320 @@ class Crystal::Call
+
+    raise error_message, inner: inner_exception
+  end
+
+  private def check_arguments_already_specified(call_errors, owner, defs, arg_types, inner_exception)
+    return unless call_errors.all?(ArgumentsAlreadySpecified)
+
+    call_errors = call_errors.map &.as(ArgumentsAlreadySpecified)
+    all_names = call_errors.flat_map(&.names).uniq
+    all_names_in_all_overloads = all_names.select do |name|
+      call_errors.all? &.names.includes?(name)
+    end
+    return if all_names_in_all_overloads.empty?
+
+    error_message = String.build do |str|
+      quoted_names = all_names.map { |name| "'#{name}'" }
+
+      if quoted_names.size == 1
+        str << "argument for parameter #{quoted_names.first} already specified"
+      else
+        str << "arguments for parameters #{quoted_names.join ", "} already specified"
+      end
+
+      str.puts
+      str.puts
+      str << "Overloads are:"
+      append_matches(defs, arg_types, str)
+    end
+
+    raise error_message, inner: inner_exception
+  end
+
+  private def check_wrong_number_of_arguments(call_errors, owner, defs, def_name, arg_types, named_args_types, inner_exception)
+    return unless call_errors.all?(WrongNumberOfArguments)
+
+    raise_matches_not_found_named_args(owner, def_name, defs, arg_types, named_args_types, inner_exception)
+  end
+
+  private def check_arguments_type_mismatch(call_errors, owner, defs, def_name, arg_types, named_args_types, inner_exception)
+    call_errors = call_errors.select(ArgumentsTypeMismatch)
+    return if call_errors.empty?
+
+    call_errors = call_errors.map &.as(ArgumentsTypeMismatch)
+    argument_type_mismatches = call_errors.flat_map(&.errors)
+
+    all_indexes_or_names = argument_type_mismatches.map(&.index_or_name).uniq
+    indexes_or_names_in_all_overloads = all_indexes_or_names.select do |index_or_name|
+      call_errors.all? &.errors.any? &.index_or_name.==(index_or_name)
+    end
+
+    return if indexes_or_names_in_all_overloads.empty?
+
+    # Only show the first error. We'll still list all overloads.
+    index_or_name = indexes_or_names_in_all_overloads.first
+
+    mismatches = argument_type_mismatches.select(&.index_or_name.==(index_or_name))
+    expected_types = mismatches.map(&.expected_type).uniq.sort_by(&.to_s)
+    actual_type = mismatches.first.actual_type
+
+    arg =
+      case index_or_name
+      in Int32
+        args[index_or_name]?
+      in String
+        named_args.try &.find(&.name.==(index_or_name))
+      end
+
+    error_message = String.build do |str|
+      argument_name =
+        case index_or_name
+        in Int32
+          case index_or_name
+          when 0 then "first argument"
+          when 1 then "second argument"
+          when 2 then "third argument"
+          when 3 then "fourth argument"
+          when 4 then "fifth argument"
+          when 5 then "sixth argument"
+          when 6 then "seventh argument"
+          else        "argument #{index_or_name + 1}"
+          end
```

Comment:
I'd prefer using `argument ##{index_or_name + 1}` in all cases.

Reply:
Yeah I think that makes more sense. Words may read a bit better. But numbers are more concise and it's a consistent style.

Reply:
Is #1 the first or second argument. There's the issue of zero being the first thing in computers. But I don't have a strong preference.

Reply:
Oh, that's a good point. 🤔

Reply:
There is already precedence in other error messages:

```crystal
def foo(& : Int32 ->)
  yield "" # Error: argument #1 of yield expected to be Int32, not String
end

foo { }

lib LibFoo
  fun foo(Int32)
end

LibFoo.foo 'x' # Error: argument #1 of 'LibFoo#foo' must be Int32, not Char
```

Reply:
Cool! I'll change it later (probably on Monday)

Reply:
Fixed!